ledger/participant-integration-api/src/main/scala/platform/store/backend/common/EventReaderQueries.scala

```diff
@@ -0,0 +1,165 @@
+  ): Vector[EventStorageBackend.Entry[Raw.TreeEvent]] = {
+
+    val witnessesColumn = "tree_event_witnesses"
+
+    val tables = List(
+      SelectTable(
+        tableName = "participant_events_create",
+        selectColumns =
+          s"$selectColumnsForTransactionTreeCreate, ${queryStrategy.constBooleanSelect(false)} as exercise_consuming",
+      ),
+      SelectTable(
+        tableName = "participant_events_consuming_exercise",
+        selectColumns =
+          s"$selectColumnsForTransactionTreeExercise, ${queryStrategy.constBooleanSelect(true)} as exercise_consuming",
+      ),
+    )
+
+    def selectFrom(tableName: String, selectColumns: String) =
+      cSQL"""
+      (
+        SELECT
+          #$selectColumns,
+          event_witnesses,
+          command_id
+        FROM
+        (
+            SELECT
+              #$selectColumns,
+              #$witnessesColumn as event_witnesses,
+              e.command_id
+            FROM
+              #$tableName e
+            JOIN parameters p
+            ON
+              p.participant_pruned_up_to_inclusive IS NULL
+              OR
+              e.event_offset > p.participant_pruned_up_to_inclusive
+            WHERE
+              e.contract_id = ${contractId.coid}
+            ORDER BY
+              e.event_sequential_id
+        ) x
+      )
+    """
+
+    val unionQuery = tables
+      .map(table =>
+        selectFrom(
+          tableName = table.tableName,
+          selectColumns = table.selectColumns,
+        )
+      )
+      .mkComposite("", " UNION ALL", "")
+
+    val query = SQL"""$unionQuery ORDER BY event_sequential_id"""
+
+    query.asVectorOf(treeEventParser(requestingParties))(connection)
+
+  }
+
+  private def fetchSeqIdOfPrecedingCreate(keyHash: String, startExclusive: EventSequentialId)(
+      connection: Connection
+  ): Option[EventSequentialId] = {
+    // As we are working in the preceding period include the start-exclusive
+    val parser = SqlParser.get[EventSequentialId](1).?
+    val query = SQL"""
+      SELECT max(event_sequential_id)
+      FROM participant_events_create
+      WHERE create_key_hash = $keyHash
+      AND event_sequential_id <= $startExclusive
+    """
+    query.as(parser.single)(connection)
+  }
+
+  def fetchContractKeyEvents(
+      keyHash: String,
+      requestingParties: Set[Party],
+      maxEvents: Int,
+      startExclusive: EventSequentialId,
+      endInclusive: EventSequentialId,
+  )(
+      connection: Connection
+  ): Vector[EventStorageBackend.Entry[Raw.TreeEvent]] = {
+
+    val precedingCreateSeq = fetchSeqIdOfPrecedingCreate(keyHash, startExclusive)(connection)
+      .getOrElse(EventSequentialId.beforeBegin)
+
+    val query = SQL"""
+        WITH key_contract AS (
+            SELECT e.contract_id cid
+            FROM participant_events_create e
+            JOIN parameters p ON p.participant_pruned_up_to_inclusive IS NULL OR event_offset > p.participant_pruned_up_to_inclusive
```

Comment:
As above, the relevant create pertaining to a given contract key might linger in the distant past beyond the pruning offset

Reply:
Additional tests to be added to `ParticipantPruningIT`